### PR TITLE
Feature/28 refactor conditional holidays

### PR DESCRIPTION
**Pull request by ckunki**

### CODE REVIEW

**Comment by ckunki** [comment]
Hello @kaklakariada, this pull request contains the changes from #25 but uses builders for conditions.

File: src/main/java/org/itsallcode/holidays/calculator/logic/conditions/builder/ConditionBuilder.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		}
+	}
+
+	private boolean negated = false;
+	private DayOfWeek[] daysOfWeek;
+	private MonthDay pivotDate;
+
+	public Condition build() {
+		if (pivotDate == null) {
+			throw new UnspecifiedPivotDateException("Cannot build DayOfWeekCondition with unspecified pivot date.");
```
Comment (kaklakariada) [inline src/main/java/org/itsallcode/holidays/calculator/logic/conditions/builder/ConditionBuilder.java]: You could also use an IllegalStateException if you want

File: src/main/java/org/itsallcode/holidays/calculator/logic/conditions/NegatedCondition.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+package org.itsallcode.holidays.calculator.logic.conditions;
+
+import java.time.Year;
+
+public class NegatedCondition extends Condition {
+
+	private final Condition other;
+
+	public NegatedCondition(Condition other) {
+		super();
```
Comment (kaklakariada) [inline src/main/java/org/itsallcode/holidays/calculator/logic/conditions/NegatedCondition.java]:
```suggestion
```

File: src/main/java/org/itsallcode/holidays/calculator/logic/parser/matcher/Patterns.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	static final Pattern ALTERNATIVE_DATE_HOLIDAY_NEGATED_DAY_OF_WEEK = buildRegexp(
+			CATEGORY, "either", MONTH, DAY, "or", "if", "not",
+			PIVOT_DAYS_OF_WEEK, "then", "fixed", MONTH_2, DAY_2, HOLIDAY_NAME);
+
+	static final Pattern FLOATING_HOLIDAY = buildRegexp( //
+			CATEGORY, "float", POSITIVE_OFFSET, DAY_OF_WEEK, DIRECTION, MONTH, DAY_OR_DEFAULT, HOLIDAY_NAME);
+
+//	static final Pattern FLOATING_HOLIDAY_WITH_OFFSET_IN_DAYS_OLD = buildRegexp( //
+//			CATEGORY, "float", POSITIVE_OFFSET, DAY_OF_WEEK, DIRECTION, MONTH, DAY_OR_DEFAULT,
+//			ADDITIONAL_OFFSET, HOLIDAY_NAME);
```
Comment (kaklakariada) [inline src/main/java/org/itsallcode/holidays/calculator/logic/parser/matcher/Patterns.java]:
```suggestion
```

File: src/main/java/org/itsallcode/holidays/calculator/logic/parser/matcher/Patterns.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	private static final String OFFSET_REGEXP = "\\d\\d?";
+	private static final String DIRECTION_REGEXP = "before|after";
+
+	// tokens for regular expressions
+	private static final Token CATEGORY = new Token(CATEGORY_GROUP, "\\S+");
+	private static final Token MONTH = new Token(MONTH_GROUP, MONTH_REGEX);
+	private static final Token DAY = new Token(DAY_GROUP, DAY_REGEX);
+	private static final Token DAY_OR_DEFAULT = new Token(DAY_GROUP, DAY.pattern + "|" + LAST_DAY);
+	private static final Token OFFSET = new Token(OFFSET_GROUP, "[+-]?\\d\\d?");
+//	private static final Token ADDITIONAL_OFFSET = new Token(ADDITIONAL_OFFSET_GROUP, OFFSET_REGEXP);
```
Comment (kaklakariada) [inline src/main/java/org/itsallcode/holidays/calculator/logic/parser/matcher/Patterns.java]:
```suggestion
```

File: src/test/java/org/itsallcode/holidays/calculator/logic/HolidayCalculationTest.java [hunk 5]
```diff
@@ -22,40 +22,189 @@
...
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.CsvSource;
 
 class HolidayCalculationTest {
 
-	AbbreviationParser<DayOfWeek> dayOfWeekParser = new AbbreviationParser<>(DayOfWeek.class);
+	static final AbbreviationParser<DayOfWeek> DAY_OF_WEEK_PARSER = new AbbreviationParser<>(DayOfWeek.class);
+
+//	static final Holiday BANK_HOLIDAY_DEC_27_OLD;
```
Comment (kaklakariada) [inline src/test/java/org/itsallcode/holidays/calculator/logic/HolidayCalculationTest.java]:
```suggestion
```

**Review by kaklakariada — APPROVED** [review]

File: src/main/java/org/itsallcode/holidays/calculator/logic/conditions/builder/ConditionBuilder.java (hunk 1 above)
Comment (ckunki) [reply src/main/java/org/itsallcode/holidays/calculator/logic/conditions/builder/ConditionBuilder.java]: 👍

Comment (ckunki) [reply src/main/java/org/itsallcode/holidays/calculator/logic/conditions/builder/ConditionBuilder.java]: Done

**Pull request merged**